src/main/java/com/artipie/gem/GemInfo.java

```
@@ -62,10 +70,28 @@ public GemInfo(final Storage storage) {
     public Response response(final String line,
         final Iterable<Map.Entry<String, String>> headers,
         final Publisher<ByteBuffer> body) {
-        return connection -> connection.accept(
-            RsStatus.NOT_IMPLEMENTED,
-            new HashSet<>(),
-            Flowable.empty()
-        );
+        final Matcher matcher = PATH_PATTERN.matcher(new RequestLineFrom(line).uri().toString());
+        final Response response;
+        if (matcher.find()) {
+            final String gem = matcher.group(1);
+            final String extension = matcher.group(2);
+            Logger.info(
+                GemInfo.class,
+                "Gem info for '%s' has been requested. Extension: '%s'",
+                gem,
+                extension
+            );
+            if (extension.equals("json")) {
+                response = new RsWithStatus(RsStatus.NOT_IMPLEMENTED);
+            } else if (extension.equals("yml")) {
+                response = new RsWithStatus(RsStatus.NOT_IMPLEMENTED);
+            } else {
+                response = new RsWithStatus(RsStatus.INTERNAL_ERROR);
+            }
+        } else {
+            response =  new RsWithStatus(RsStatus.INTERNAL_ERROR);
```

@Sammers21 it should be 404. This is not an internal error

---

Disagree. If this case is executed, the requested is routed in the wrong way = internal error.

---

@Sammers21 your solution is not by rest practices. This is not an error that is caused on the server side, this is simply fetching of a resource that is not supported by the app, or doesnt exist.

---

This is an error on the server side: request routed in the wrong way.

---

@Sammers21 this is not unexpected behavior on server side it is perfectly valid to request the invalid path. I wont argue with you any more.
@g4s8 What is your opinion?

---

@Vatavuk , You did not get it. The regexp pattern for the route and inside the route is the same. The method would not be called in case request is valid. If the method is called, the request is necessarily valid.

---

@Vatavuk @Sammers21 if this condition should never be matched because of routing rules, I'd throw an exception here

---

@Sammers21 let's throw exception than and  we can merge, from my side

---

@Vatavuk , Done